### PR TITLE
Automated automated testing for CNN pipeline

### LINKED ISSUE #127: Adjust cnn scraper test suite to use test data
As part of #125 we will probably need / want to adjust the way the unit tests are written as well, and Instead of having a long list of manually populated we should create test data which represents the "state" at different points of the pipeline.

This way our tests of each pipeline utility method can iterate through the data that has the correct format for each step.

This also will fix the situation where discovering a new edge case would require a whole bunch of duplicate test code across different (similar) components (e.g. "getNameFromAttribution" and "getAffiliationFromAttribution")

### PR DESCRIPTION
**Pull request by slifty**
This PR creates a mechanism (and utility) for us to easily expand the CNN test suite to include real world data as we find instances of the system failing.

For documentation on how to use these tools meaningfully go check the new `docs/TESTING.md` which is part of this PR.

run `yarn generate:test:cnn -u [FULL_TRANSCRIPT_URL_HERE]`

Resolves #127 
Related to #47 since we technically now have TESTING.md

### CODE REVIEW

File: src/scripts/generateCnnUtilTest.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  normalizeStatementSpeakers,
+  removeNetworkAffiliatedStatements,
+  removeUnattributableStatements,
+} from '../server/utils/cnn'
+import logger from '../server/utils/logger'
+
+const FILE_ENCODING = 'utf8'
+const BASE_PATH = `${__dirname}/../server/utils/__test__/data/cnn`
+
+const scrapeSequence = [
```
Comment (slifty) [inline src/scripts/generateCnnUtilTest.js]: @reefdog what do you think about me updating the CnnStatementScraper to have its sequence actually be a class variable, so this could be scrapeSequence = CnnStatementScraper.scrapeSequence

That way if the sequence changes we don't need to update that change in three places...

File: src/scripts/generateCnnUtilTest.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+const createTestSuiteDirectory = (suiteIndex) => {
+  const dirPath = getTestSuiteDirectoryPath(suiteIndex)
+  if (!fs.existsSync(dirPath)) {
+    fs.mkdirSync(dirPath)
+  }
+}
+
+const saveTestUrl = (suiteIndex, url) => {
+  fs.writeFileSync(
```
Comment (reefdog) [inline src/scripts/generateCnnUtilTest.js]: Worth wrapping in a try/catch.

File: src/scripts/generateCnnUtilTest.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    .map(fileName => parseInt(fileName, 10))
+    .filter(Number.isInteger)
+    .sort((a, b) => a - b)
+  return testSuiteIndexes.length === 0 ? 1 : testSuiteIndexes.slice(-1)[0] + 1
+}
+
+const createTestSuiteDirectory = (suiteIndex) => {
+  const dirPath = getTestSuiteDirectoryPath(suiteIndex)
+  if (!fs.existsSync(dirPath)) {
+    fs.mkdirSync(dirPath)
```
Comment (reefdog) [inline src/scripts/generateCnnUtilTest.js]: Worth wrapping in a try/catch. (All filesystem changes should be IMO.)

File: src/scripts/generateCnnUtilTest.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  fs.writeFileSync(
+    getTestSuiteUrlPath(suiteIndex),
+    url,
+    { encoding: FILE_ENCODING },
+  )
+}
+
+const saveTestStep = (suiteIndex, stepIndex, name, input, output) => {
+  const testStepContent = {
+    name,
```
Comment (reefdog) [inline src/scripts/generateCnnUtilTest.js]: It's recommended we not use `name`; it's not technically a reserved word, but it is on the list of common properties to avoid using as constant names.

File: src/server/utils/__test__/data/cnn/index.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+const testSuiteBasePath = `${__dirname}`
+const testSuites = {}
+
+const getFunctionNameFromFileName = fileName => path
+  .parse(fileName).name
+  .split('_')[1]
+
+const loadTestSuite = suitePath => fs
+  .readdirSync(suitePath)
+  .filter(fileName => fileName.endsWith('.json'))
```
Comment (reefdog) [inline src/server/utils/__test__/data/cnn/index.js]: Wanna move this to an `isJsonFile()` utility function (in this file)?

File: src/scripts/generateCnnUtilTest.js [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    return null
+  }
+}
+
+const getNextTestSuiteIndex = () => {
+  const testSuiteIndexes = fs
+    .readdirSync(BASE_PATH)
+    .slice(0, -1)
+    .map(fileName => parseInt(fileName, 10))
+    .filter(Number.isInteger)
```
Comment (reefdog) [inline src/scripts/generateCnnUtilTest.js]: It's not quite clear to me why, if you're specifically parsing them as integers one line up, you'd need to filter them to only integers. Is that to get rid of `NaN`s or other weirdness that slip through?

File: src/scripts/generateCnnUtilTest.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+const getUrlFromTestSuite = (suiteIndex) => {
+  try {
+    return fs.readFileSync(getTestSuiteUrlPath(suiteIndex), FILE_ENCODING)
+  } catch {
+    return null
+  }
+}
+
+const getNextTestSuiteIndex = () => {
```
Comment (reefdog) [inline src/scripts/generateCnnUtilTest.js]: The body of this function is not super legible, but I also don't think you should expand the code too much to be self-documenting. Can you just add a note about how this function does what it does? The other functions in this file are more legible and straightforward, this one feels a bit magical (despite the obvious name).

File: src/scripts/generateCnnUtilTest.js [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+const scrapeRawTranscript = (url) => {
+  const scraper = new GenericScraper(url)
+  return scraper.run()
+}
+
+const testSuiteIndex = getTestSuiteIndex()
+const currentIndex = testSuiteIndex || getNextTestSuiteIndex()
+const currentUrl = getUrl() || getUrlFromTestSuite(currentIndex)
+
+if (!currentUrl) {
```
Comment (reefdog) [inline src/scripts/generateCnnUtilTest.js]: We have `isTranscriptUrl()`; we should use that here to avoid head-scratching situations like the one in the demo.

File: src/scripts/generateCnnUtilTest.js [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+  logger.error('Pass in a valid CNN transcript URL with the `-u` parameter, or a suite index to re-run with the `-i` parameter')
+  process.exit()
+}
+
+createTestSuiteDirectory(currentIndex)
+saveTestUrl(currentIndex, currentUrl)
+scrapeRawTranscript(currentUrl)
+  .catch(error => logger.error(`Scraping failed. ${error}`))
+  .then(scrapeResult => generateTestSuiteFromSequence(currentIndex, scrapeSequence, scrapeResult))
+  .finally(() => process.exit())
```
[the remark this answers is not in the document]
Comment (slifty) [reply src/scripts/generateCnnUtilTest.js]: Yes.

File: src/server/workers/scrapers/CnnTranscriptStatementScraper.js [hunk 10]
```diff
@@ -97,7 +88,7 @@ class CnnTranscriptStatementScraper extends AbstractStatementScraper {
   }
 
   statementScrapeHandler = (responseString) => {
-    const transcript = this.getTranscriptText(responseString)
+    const transcript = getTranscriptTextFromHtml(responseString)
```
[the remark this answers is not in the document]
Comment (slifty) [reply src/server/workers/scrapers/CnnTranscriptStatementScraper.js]: Yes indeed!  I'll go forward with that

File: src/scripts/generateCnnUtilTest.js (hunk 4 above)
Comment (slifty) [reply src/scripts/generateCnnUtilTest.js]: YOU'RE not technically a reserved word

File: src/scripts/generateCnnUtilTest.js (hunk 6 above)
Comment (slifty) [reply src/scripts/generateCnnUtilTest.js]: there were NaN's -- .DS_Store, _url.txt, and in theory if there was ever a .json file with a string character it would break stuff.  This is defensive programming at this point since the main culprits are filtered earlier I believe.

File: src/scripts/generateCnnUtilTest.js (hunk 4 above)
Comment (reefdog) [reply src/scripts/generateCnnUtilTest.js]: Good point, change it to `Justin`.

File: src/scripts/generateCnnUtilTest.js [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+      { encoding: FILE_ENCODING },
+    )
+  } catch (e) {
+    logger.error(e)
+  }
+}
+
+const saveTestStep = (suiteIndex, stepIndex, stepName, input, output) => {
+  const testStepContent = {
+    name: stepName,
```
Comment (reefdog) [inline src/scripts/generateCnnUtilTest.js]: So 😬 I think this is still a best practices violation, to have a property we made up be named `name`. This could be called anything, right? Can we call it even `fnName` or something? (Or keep it `stepName`?)
